lib/msal-core/src/UserAgentApplication.ts

```diff
@@ -2149,39 +2149,21 @@ export class UserAgentApplication {
      * Utils function to create the Authentication
      * @param {@link account} account object
      * @param scopes
-     * @param silentCall
      *
-     * @returns {string} token type: id_token or access_token
+     * @returns {string} token type: token, id_token or id_token token
      *
      */
-    private getTokenType(accountObject: Account, scopes: string[], silentCall: boolean): string {
-        /*
-         * if account is passed and matches the account object/or set to getAccount() from cache
-         * if client-id is passed as scope, get id_token else token/id_token_token (in case no session exists)
-         */
-        let tokenType: string;
-
-        // acquireTokenSilent
-        if (silentCall) {
-            if (Account.compareAccounts(accountObject, this.getAccount())) {
-                tokenType = (scopes.indexOf(this.config.auth.clientId) > -1) ? ResponseTypes.id_token : ResponseTypes.token;
-            }
-            else {
-                tokenType  = (scopes.indexOf(this.config.auth.clientId) > -1) ? ResponseTypes.id_token : ResponseTypes.id_token_token;
-            }
-            return tokenType;
-        }
-        // all other cases
-        else {
-            if (!Account.compareAccounts(accountObject, this.getAccount())) {
-                tokenType = ResponseTypes.id_token_token;
-            }
-            else {
-                tokenType = (scopes.indexOf(this.clientId) > -1) ? ResponseTypes.id_token : ResponseTypes.token;
-            }
-            return tokenType;
+    private getTokenType(accountObject: Account, scopes: string[]): string {
+        // Returns response type "token" if request has no login scopes and the account passed matches the current account object
+        if (!ScopeSet.isLoginScopes(scopes, this.clientId) && Account.compareAccounts(accountObject, this.getAccount())) {
+            return ResponseTypes.token;
         }
 
+        /**
+         * Returns response type "id_token token" if both login scopes and access token scopes are present
+         * Returns response type "id_token" if only login scopes are present
+         */
+        return (scopes.length > 2) ? ResponseTypes.id_token_token : ResponseTypes.id_token;
```

Comment:
This assumes `openid` and `profile` are always included, correct? I would add a comment making that clear (or calculating the length without those two scopes).

Reply:
Correct, I've made sure any login scopes will already be in the array by the time this method even gets called. I'll add the comment.

Reply:
Nevermind, I'm unsure about all cases being covered here. I'll spend some more time making sure this is right and fixing it if not. Thanks @jasonnutter!

Reply:
@jasonnutter I did a lot of work to summarize the response types table in the doc, there was some repetition. After that, I made sure every single case was covered by the code (it wasn't). The new code, although a little more verbose, covers every case in what I believe to be the least amount of checks possible. I also added comments explaining the logic to make sure the intention isn't lost in the future (making sure to answer "why" over "what". Thanks again for catching this, I think I'll need to add some tests for this behavior at the public APIs level.